src/rp2_common/hardware_uart/uart.c

```diff
@@ -69,6 +70,65 @@ void uart_deinit(uart_inst_t *uart) {
     uart_reset(uart);
 }
 
+static uint32_t uart_disable_before_lcr_write(uart_inst_t *uart) {
+    // Notes from PL011 reference manual:
+    //
+    // - Before writing LCR, must disable UART and wait for current TX + RX
+    //   char to finish
+    //
+    // - There is a BUSY flag which waits for the current TX char, but this is
+    //   OR'd with TX FIFO !FULL, so not usable when FIFOs are enabled and
+    //   potentially nonempty
+    //
+    // - FIFOs can't be set to disabled whilst a character is in progress
+    //   (else "FIFO integrity is not guaranteed")
+    //
+    // Combination of these means there is no general way to halt and poll for
+    // end of TX character, if FIFOs may be enabled. Either way, there is no
+    // way to poll for end of RX character.
+    //
+    // So... we're going to insert a 15 baud period delay before changing
+    // settings (comfortably higher than max data + start + stop + parity).
+    // Anything else would require API changes to permit a non-enabled UART
+    // state after init() where settings can be changed safely.
+
+    uint32_t cr_save = uart_get_hw(uart)->cr;
+    hw_clear_bits(&uart_get_hw(uart)->cr,
+        UART_UARTCR_UARTEN_BITS | UART_UARTCR_TXE_BITS | UART_UARTCR_RXE_BITS);
+
+    uint32_t current_ibrd = uart_get_hw(uart)->ibrd;
+    uint32_t current_fbrd = uart_get_hw(uart)->fbrd;
+
+    // Note: Maximise precision here. Show working, the compiler will mop this up.
+    // Create a 16.6 fixed-point fractional division ratio; then scale to 32-bits.
+    uint32_t brdiv_ratio = 64u * current_ibrd + current_fbrd;
+    brdiv_ratio <<= 10;
+    // 3662 is ~(15 * 244.14) where 244.14 is 16e6 / 2^16
+    uint32_t scaled_freq = clock_get_hz(clk_peri) / 3662ul;
+    uint32_t wait_time_us = brdiv_ratio / scaled_freq;
+    busy_wait_us(wait_time_us);
+
+    return cr_save;
+}
+
+static void uart_write_lcr_bits_masked(uart_inst_t *uart, uint32_t values, uint32_t write_mask) {
+    invalid_params_if(UART, uart != uart0 && uart != uart1);
+    uint32_t cr_save;
+    bool enabled = uart_is_enabled(uart);
```

Comment:
I mentioned this in a previous review, but i would like to get rid of the `uart_is_enabled` checks in the caller, and just put it in the uart_disable_before_lcr_write (which can just do the delay if the UART was enabled based on its read of the CR).

it is then fine to unconditionally set the CR at the bottom of the callers too.

Reply:
`uart_disable_before_lcr_write` is only called from one place, so I guess an alternative might be to just fold that code back inside `uart_write_lcr_bits_masked` ? :shrug:

Reply:
@kilograham I've just pushed a fix
@lurch We could roll up those two functions but I don't think it will make a difference to the object; the compiler will inline them because (as you observe) `uart_disable_before_lcr_write` is only called from one place.
I think it is more readable as it is - the intricacies of the UART disabling are removed from the functional code in `uart_write_lcr_bits_masked`